docs/guides/go-built-in.md

```diff
@@ -311,112 +310,95 @@ information on why the transaction was rejected.
 
-```go
-func (app *KVStoreApplication) BeginBlock(req abcitypes.RequestBeginBlock) abcitypes.ResponseBeginBlock {
-	app.onGoingBlock = app.db.NewTransaction(true)
-	return abcitypes.ResponseBeginBlock{}
-}
-```
+First, let's create a new Badger transaction during `FinalizeBlock`. All application transactions in the current block will be executed within this Badger transaction.
+Next, let's modify `FinalizeBlock` to add the `key` and `value` to the database transaction every time our application receives a new application transaction through `RequestFinalizeBlock`.
 
-Next, let's modify `DeliverTx` to add the `key` and `value` to the database transaction every time our application
-receives a new application transaction through `RequestDeliverTx`.
+Note that we check the validity of the transaction _again_ during `FinalizeBlock`.
 
 ```go
-func (app *KVStoreApplication) DeliverTx(req abcitypes.RequestDeliverTx) abcitypes.ResponseDeliverTx {
-	if code := app.isValid(req.Tx); code != 0 {
-		return abcitypes.ResponseDeliverTx{Code: code}
-	}
-
-	parts := bytes.SplitN(req.Tx, []byte("="), 2)
-	key, value := parts[0], parts[1]
-
-	if err := app.onGoingBlock.Set(key, value); err != nil {
-		log.Panicf("Error writing to database, unable to execute tx: %v", err)
-	}
-
-	return abcitypes.ResponseDeliverTx{Code: 0}
+func (app *KVStoreApplication) FinalizeBlock(_ context.Context, req *abcitypes.RequestFinalizeBlock) (*abcitypes.ResponseFinalizeBlock, error) {
+    var txs = make([]*abcitypes.ExecTxResult, len(req.Txs))
+
+    app.onGoingBlock = app.db.NewTransaction(true)
+    for i, tx := range req.Txs {
+        if code := app.isValid(tx); code != 0 {
+            log.Printf("Error: invalid transaction index %v", i)
+            txs[i] = &abcitypes.ExecTxResult{Code: code}
+        } else {
+            parts := bytes.SplitN(tx, []byte("="), 2)
+            key, value := parts[0], parts[1]
+            log.Printf("Adding key %s with value %s", key, value)
+
+            if err := app.onGoingBlock.Set(key, value); err != nil {
+                log.Panicf("Error writing to database, unable to execute tx: %v", err)
+            }
+
+            log.Printf("Successfully added key %s with value %s", key, value)
+
+            // Marshal the ExecTxResult to the txs slice element
+            txs[i] = &abcitypes.ExecTxResult{}
+        }
+    }
+
+    return &abcitypes.ResponseFinalizeBlock{
+    TxResults:        txs,
+    }, nil
 }
 ```
 
-Note that we check the validity of the transaction _again_ during `DeliverTx`.
-Transactions are not guaranteed to be valid when they are delivered to an
-application, even if they were valid when they were proposed.
-This can happen if the application state is used to determine transaction
-validity. Application state may have changed between the initial execution of `CheckTx`
```

Comment:
this comment is still relevant, once adapted to finalizeBlock

Reply:
I made a small suggestion in the explantion of FinalizeBlock to address this.

Reply:
@lasarojc - the text should read something like this then?


Transactions are not guaranteed to be valid when they are delivered to an application, even if they were valid when they were proposed.

This can happen if the application state is used to determine transaction
validity. The application state may have changed between the initial execution of `CheckTx` and the transaction delivery in `FinalizeBlock` in a way that rendered the transaction no longer valid.

Note that `FinalizeBlock` **cannot** yet commit the Badger transaction we were building during the block execution.

 Other methods, such as `Query`, rely on a consistent view of the application's
 state, the application should only update its state by committing the Badger transactions when the full block has been delivered and the `Commit` method is invoked.

Reply:
Looks great!